File: 2019/tourism.py
# part from https://github.com/elibroftw/contest-questions/blob/master/CCC/2019/S4%20Tourism.py

import math

N, K = [int(char) for char in input().split()]
A = [int(char) for char in input().split()]

# N = 5
# K = 3
# A = [2, 5, 7, 1, 4]

# n/k remainder is 0, that means that at each day, move k steps
# else you have leeway of k - remainder

dictionary = {}


def recurse(arr, n, k):

    tuple_arr = tuple(arr)

    if (n, k, tuple_arr) in dictionary:
        return dictionary[(n, k, tuple_arr)]
    else:
        if n <= k:
            temp = max(arr)
            dictionary[(n, k, tuple_arr)] = temp
            return temp

        remainder = n % k

        if remainder == 0:
            scores = 0
            for i in range(0, n, k):
                scores += max(arr[i:i + k])
            dictionary[(n, k, tuple_arr)] = scores
            return scores
        else:
            possible_max = 0
            for i in range(remainder, k + 1):
                today_max = max(arr[:i])

                future_max = recurse(arr[i:], n - i, k)

                possible_max = max(possible_max, today_max + future_max)

            dictionary[(n, k, tuple_arr)] = possible_max
            return possible_max


out = recurse(A, N, K)
print(out)

# def recurse(arr, sofar):

#     if sofar > days:
#         return 0

#     if len(arr) == 0:
#         return 0
#     elif len(arr) == 1:
#         return arr[0]
#     else:
#         possible_futures = []
#         for i in range(1, min(K + 1, len(arr) + 1)):
#             subset = arr[:i]

#             max_today = max(subset)

#             max_future = recurse(arr[i:], sofar + 1)

#             possible_futures.append(max_today + max_future)

#         future = max(possible_futures)
#         return future


# out = recurse(A[:], 1)
# print(out)

File: 2019/tourism.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <map>

using namespace std;

std::map<std::vector<int>, int> dict;

int recurse(std::vector<int> arr, int n, int k)
{
    std::vector<int> tempvec(arr);
    tempvec.push_back(n);
    tempvec.push_back(k);

    if (dict.find(tempvec) == dict.end())
    {
        if (n <= k)
        {
            int temp = *max_element(arr.begin(), arr.end());
            dict[tempvec] = temp;
            return temp;
        }

        int remainder = n % k;

        if (remainder == 0)
        {
            int scores = 0;

            int i = 0;
            while (i < n)
            {
                int temp = *max_element(arr.begin() + i, arr.begin() + i + k);
                scores += temp;

                i += k;
            }

            dict[tempvec] = scores;
            return scores;
        }
        else
        {
            int possible_max = 0;
            for (int i = remainder; i < k + 1; ++i)
            {
                int today_max = *max_element(arr.begin(), arr.begin() + i);

                std::vector<int> temp_vec = std::vector<int>(arr.begin() + i, arr.end());
                std::vector<int> temp2(temp_vec);
                int future_max = recurse(temp2, n - i, k);

                possible_max = std::max(possible_max, today_max + future_max);
            }

            dict[tempvec] = possible_max;
            return possible_max;
        }
    }
    else
    {
        return dict[tempvec];
    }
}

int main()
{

    int N;
    int K;

    std::vector<int> A;

    cin >> N >> K;

    for (int i = 0; i < N; ++i)
    {
        int temp;
        cin >> temp;
        A.push_back(temp);
    }

    int out = recurse(A, N, K);

    cout << out << endl;

    return 0;
}